crates/lsp/src/main_loop.rs

```diff
@@ -360,15 +319,17 @@ impl GlobalState {
                             // Currently ignored
                         },
                         LspNotification::DidCloseTextDocument(params) => {
-                            handlers_state::did_close(params, &mut self.world, &self.auxiliary_event_tx)?;
+                            handlers_state::did_close(params, &mut self.world)?;
                         },
                     }
                 },
 
                 LspMessage::Request(request, tx) => {
                     match request {
                         LspRequest::Initialize(params) => {
-                            respond(tx, handlers_state::initialize(params, &mut self.lsp_state, &mut self.world), LspResponse::Initialize)?;
+                            // Unwrap: `Initialize` method should only be called once.
+                            let log_tx = self.log_tx.take().unwrap();
+                            respond(tx, handlers_state::initialize(params, &mut self.lsp_state, &mut self.world, log_tx), LspResponse::Initialize)?;
```

Comment:
I quite like how lsp_server works here. Instead of `Initialize` being an event you handle in your main loop, it is instead handled through explicit calls to `initialize_start()` to receive the `Initialize` request, and `initialize_finish()` to send back an `InitializeResult`. 

I like that because it really drives home that they are once-per-session methods, not ones that can be recalled through the main loop.

https://docs.rs/lsp-server/latest/lsp_server/struct.Connection.html#method.initialize_start